### PR TITLE
Major `credit.postblock` update + major `credit.physics_core` update + major bugfix + minor code optimization and cleaning

### PR DESCRIPTION
**Pull request by yingkaisha**
**New feature**
The new branch with `credit.physic_core` and `credit.physic_constants` that works for global models. It will be developed based on ERA5 pressure level data. `credit.postblock` is added to enable various post-processing-related operations. The following features will be added:

* New `parser.py` workflow that pulls channel indices based on given variable names. This will be part of the `credit.postblock` integrations down the road. `example.yml` and `example_physics_single.yml` have been updated based on the changes in the new parser.
* `credit.parser` is also updated for the use of mean, std Pytorch tensors in `Normalize_ERA5_and_Forcing`
* `Normalize_ERA5_and_Forcing` in `transform.py` can now inverse transform x in `inverse_transform_input`
* Negative tracer fixer in `postblock.py` is ready for use. It converts values below a threshold to the threshold value.
* SST forcing with dynamic ocean masking is ready for single-step training. `train.py` has been updated for this feature (`train_multistep.py` has NOT been updated yet; some multi-step issues will be addressed together in a new PR).
* All new features above do not interrupt the current usage of CREDIT

**Bugfix**
* In `transforms.py`, `ToTensor_ERA5_and_Forcing` now always produces `float32` PyTorch tensors. Thus, other `.float()` castings in `trainerERA5_v2` and `trainerERA5_multistep_grad_accum` were commented out.
* Fixed the handling of `'x_forcing_static'` channel=1 cases in `ToTensor_ERA5_and_Forcing`, `unsqueeze` is now assigned on the correct dimension.
* `Normalize_ERA5_and_Forcing` now loads the entire `xr.dataset` of mean and std values into memory for `transform`. It now uses PyTorch tensors directly for `transform_array`, `inverse_transform`, and `inverse_transform_input`.
* The multi-step iteration in `trainerERA5_v2` is wrong. This feature has been removed because `trainerERA5_v2` is prepared for single-step training only. asserts were added to verify `forecast_len=0`
* Fixed FuXi patch number calculation:
```
# wrong: 
input_resolution = int(img_size[1] / patch_size[1] / 2), int(img_size[2] / patch_size[2] / 2)
# correct:
input_resolution = round(img_size[1] / patch_size[1] / 2), round(img_size[2] / patch_size[2] / 2)
```
* Fixed `credit.parser` processes on handling scheduler
* Fixed `credit.loss` and `credit.metrics` when computing loss function values for diagnostic variables
* Cleaned asserts in `ERA5_and_Forcing_Dataset` and `ERA5_and_Forcing_MultiStep`; they exist in `credit.parser` already
* Some general cleanings to remove unnecessary comments
* Some "cached" example configs were deleted because the data no longer exists in my scratch. 

**Other notes**

* `codebook` related features were commented out in `trainerERA5_v2` because they were not implemented in the new multi-step trainer. 
* The single-step trainer `trainerERA5_v2` now supports `forcast_len=0` only. This is to avoid confusion about multi-step training. Multi-step training must use its own dataset, dataloader, and trainer.
* DO NOT test diagnostic-variable-related features in multi-step training and rollout. Two new PRs will be ready to address the problems there.  
* A few things have been renamed, e.g., postBlock --> postblock. This does not impact the usage of existing features.

### CODE REVIEW

File: credit/models/fuxi.py [hunk 1]
```diff
@@ -344,8 +347,16 @@ def __init__(self,
         if self.pad_lat > 0:
             logger.info(f"Padding each pole using a reflection with {self.pad_lat} pixels")
 
+        self.use_post_block = post_conf["use_skebs"]
```
Comment (dkimpara) [inline credit/models/fuxi.py]: since you've changed the config to 
```
 post_conf: 
        activate: True
```
this line needs to be changed to reflect that

Please also update `crossformer.py` and `unet.py`

File: credit/parser.py [hunk 2]
```diff
@@ -259,16 +261,154 @@ def CREDIT_main_parser(conf, parse_training=True, parse_predict=True, print_summ
...
-        conf['model']['post_conf'] = {'use_skebs': False}
-    elif 'use_skebs' not in conf['model']['post_conf']:
-        conf['model']['post_conf']['use_skebs'] = False
-    if 'image_width' not in conf['model']['post_conf']:
-        conf['model']['post_conf']['image_width'] = conf['model']['image_width']
+    # trun-off model physcis if post_conf does not exist
+    conf['model'].setdefault('post_conf', {'activate': False, 'use_skebs': False})
+    
+    # default to True, will turn it off if no post action is needed
+    conf['model']['post_conf'].setdefault('activate', True)
```
Comment (dkimpara) [inline credit/parser.py]: these lines no longer needed with the below changes

File: credit/parser.py [hunk 3]
```diff
@@ -259,16 +261,154 @@ def CREDIT_main_parser(conf, parse_training=True, parse_predict=True, print_summ
...
     # conf['model'] section
 
-    if 'post_conf' not in conf['model']:
-        conf['model']['post_conf'] = {'use_skebs': False}
-    elif 'use_skebs' not in conf['model']['post_conf']:
-        conf['model']['post_conf']['use_skebs'] = False
-    if 'image_width' not in conf['model']['post_conf']:
-        conf['model']['post_conf']['image_width'] = conf['model']['image_width']
+    # trun-off model physcis if post_conf does not exist
+    conf['model'].setdefault('post_conf', {'activate': False, 'use_skebs': False})
```
Comment (dkimpara) [inline credit/parser.py]: I'm going to suggest a lot of changes to enchance readability and debuggability.

Let's set defaults like this:
```
conf['model'].setdefault('post_conf', {'activate': False})
conf['model']['post_conf'].setdefault('skebs', {'activate': False})
conf['model']['post_conf'].setdefault('tracer_fixer', {'activate': False})
conf['model']['post_conf'].setdefault(''mass_fixer', {'activate': False})
```
and so on

Then check for irregularities (if post_conf exists, then the previous commands does nothing). If activation fields are inconsistent we check for that here:
```
post_conf = conf['model']['post_conf']
if (post_conf['activate'] and not 
     (post_conf['skebs']['activate'] or post_conf['tracer_fixer']['activate'])
):
    raise("post conf activate is True but no post conf blocks specified
elif post_conf['use_skebs']

if (not post_conf['activate'] and 
     (post_conf['skebs']['activate'] or post_conf['tracer_fixer']['activate'])
):
    raise("post conf activate is False but post conf blocks specified

File: credit/parser.py [hunk 4]
```diff
@@ -259,16 +261,154 @@ def CREDIT_main_parser(conf, parse_training=True, parse_predict=True, print_summ
...
+    
+    # default to True, will turn it off if no post action is needed
+    conf['model']['post_conf'].setdefault('activate', True)
+    
+    if conf['model']['post_conf']['activate']:
```
Comment (dkimpara) [inline credit/parser.py:288]: This level of nested if statement is unnecessary after doing the parsing checks i suggested above. the use_post flag can also be removed because of the parsing checks. Can you please change the next section to the following logic
```
if use_skebs:
   # check for skebs defaults here
   
if use_tracer_fix:
  # set tracer defaults/fields here
 
```
and so on

File: credit/parser.py [hunk 5]
```diff
@@ -259,16 +261,154 @@ def CREDIT_main_parser(conf, parse_training=True, parse_predict=True, print_summ
...
 
-    if 'post_conf' not in conf['model']:
-        conf['model']['post_conf'] = {'use_skebs': False}
-    elif 'use_skebs' not in conf['model']['post_conf']:
-        conf['model']['post_conf']['use_skebs'] = False
-    if 'image_width' not in conf['model']['post_conf']:
-        conf['model']['post_conf']['image_width'] = conf['model']['image_width']
+    # trun-off model physcis if post_conf does not exist
+    conf['model'].setdefault('post_conf', {'activate': False, 'use_skebs': False})
+
```
Comment (dkimpara) [inline credit/parser.py:286]: lets add a copy of the model configs to post_conf:
```
conf['model']['post_conf']['mode'] = {k: v for k,v in conf['model'].items() if k != 'post_conf'} 
```

File: tests/test_postBlock.py [hunk 6]
```diff
@@ -4,7 +4,8 @@
     
 def test_SKEBS_rand():
     image_width = 100
-    conf = {"post_conf": {"use_skebs": True, "image_width": image_width}}
+    conf = {"post_conf": {"use_skebs": True, "image_width": image_width,}}
```
Comment (dkimpara) [inline tests/test_postBlock.py]: can you please write a separate unit test for the post-processing modules you're adding?


to adjust to the new config schema, we need:
```
conf = {"post_conf": 
                {"activate": True,
                "skebs": {activate: True, "image_width": image_width}
                 }
            }

```

File: credit/parser.py [hunk 7]
```diff
@@ -259,16 +261,151 @@ def CREDIT_main_parser(conf, parse_training=True, parse_predict=True, print_summ
...
+    post_conf = conf['model']['post_conf']
+    activate_any = any([post_conf[post_module]['activate'] 
+                        for post_module in post_list])
+    if post_conf['activate'] and not activate_any:
+        raise("post_conf is set activate, but no post modules specified")
+    
+    
+    if conf['model']['post_conf']['activate']:
+        # copy only model configs to post_conf subdictionary
+        conf['model']['post_conf']['model'] = {k: v for k,v in conf['model'].items() if k != 'post_conf'}
```
Comment (dkimpara) [inline credit/parser.py:290]: @yingkaisha also copy data keys
